// @apollo/client-integration-nextjs/dist/index.rsc.cjs
'use strict';

var clientReactStreaming = require('@apollo/client-react-streaming');

// src/index.shared.ts

// src/bundleInfo.ts
var bundle = {
  pkg: "@apollo/client-integration-nextjs",
  client: "ApolloClient",
  cache: "InMemoryCache"
};
var ApolloClient = class extends clientReactStreaming.ApolloClient {
  /**
   * Information about the current package and it's export names, for use in error messages.
   *
   * @internal
   */
  static info = bundle;
};
var InMemoryCache = class extends clientReactStreaming.InMemoryCache {
  /**
   * Information about the current package and it's export names, for use in error messages.
   *
   * @internal
   */
  static info = bundle;
};
exports.built_for_rsc = true;

Object.defineProperty(exports, "DebounceMultipartResponsesLink", {
  enumerable: true,
  get: function () { return clientReactStreaming.DebounceMultipartResponsesLink; }
});
Object.defineProperty(exports, "RemoveMultipartDirectivesLink", {
  enumerable: true,
  get: function () { return clientReactStreaming.RemoveMultipartDirectivesLink; }
});
Object.defineProperty(exports, "SSRMultipartLink", {
  enumerable: true,
  get: function () { return clientReactStreaming.SSRMultipartLink; }
});
Object.defineProperty(exports, "registerApolloClient", {
  enumerable: true,
  get: function () { return clientReactStreaming.registerApolloClient; }
});
exports.ApolloClient = ApolloClient;
exports.InMemoryCache = InMemoryCache;
